man/ola_rdm_get.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.37.1.
.TH OLA_RDM_GET "1" "December 2015" "ola_rdm_get  " "User Commands"
.SH NAME
ola_rdm_get \- Get the value of a PID for an RDM device
.SH SYNOPSIS
.B ola_rdm_get
\fI--universe <universe> --uid <uid> <pid> <value>\fR
.SH DESCRIPTION
Get the value of a pid for a device.
Use 'ola_rdm_get \fB\-\-list\-pids\fR' to get a list of applicable pids.
.TP
\fB\-\-frames\fR
display the raw RDM frames if available.
.TP
\fB\-\-uid\fR <uid>
the UID of the device to control.
.HP
\fB\-d\fR, \fB\-\-sub\-device\fR <device> target a particular sub device (default is 0)
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help message and exit.
.TP
\fB\-l\fR, \fB\-\-list\-pids\fR
display a list of pids that support GET.
.TP
\fB\-p\fR, \fB\-\-pid\-location\fR
the directory to read PID definitions from
.HP
\fB\-u\fR, \fB\-\-universe\fR <universe> universe number.


NEWS
x/y/2023 ola-0.10.9
 Features:
 * Python 3 compatibility across the board (including the RDM Responder Tests)!
   #1506
 * Support for the JMS USB2DMX PRO V2.1 device via the FTDI plugin #1728

 API:
 * Python: Add a fetch current DMX example.

 RDM Tests:
 * Python 3 compatibility of the RDM Tests #1599
 * Fix a longstanding bug in the GetMaxPacketSize RDM test around timeouts

 Bugs:
 * Fix some tests not working on Big Endian machines because our RPC is encoded
   with native format and those tests used sample data from a Little Endian
   machine
 * Renamed EndpointNoticationEvent(sic) to EndpointNotificationEvent in the
   E1.33 EndpointManager code
 * Check if librt exists and add it to the build if required
 * Ensure we can build on later Protobuf and C++ and silence deprecation
   messages where required
 * Fix the NetBSD build
 * Clarify the usage of some of the various Enttec Open DMX USB plugins
 * Relax the timing on the ClientWrapperTest so it doesn't fail occasionally
   #1714
 * Correct the documentation of some existing limits in the E1.31 plugin
 * Fix some undefined behaviour in TimeoutManager::ExecuteTimeouts #1726
 * Fix UARTs on Debian Bullseye #1749
 * Fix Python version detection for versions > 3.9 #1755, #1780
 * Update Angular and Angular-Route versions used by the new web UI and add
   workaround to be compatible #1782 Debian #1016447
 * Significantly reduce the delay when using StreamingClient with EPoll and
   multiple universes #1793

 Internal:
 * Add more debugging so we can tell why setting the baud rate failed
 * Replace "currentThread" with "current_thread" in Python since the former has
   been deprecated
 * Fix lots more internal typos and grammar
 * Add unit tests for more of the Python RDM Test classes
 * Sort the Python imports with isort

22/11/2020 ola-0.10.8
 Features:
 * Support DragonFly BSD
 * Add another Debian test
 * Add the E1.33 NACK Reasons
 * Add the E1.37-7 NACK Reasons
 * Use clock_gettime's CLOCK_MONOTONIC if available
 * Collect the PDL size in the model collector

 API:
 * Deprecated Clock::CurrentTime and added Clock::CurrentMonotonicTime and 
   Clock::CurrentRealTime. Occurrences of Clock::CurrentTime have been
   replaced with either variant as applicable.

 RDM Tests:
 * Fix a bug in the RDM tests to ensure they can cope with 255 personalities
   #1508
 * Only add valid network interface IDs to the test list so future tests don't
   fail
 * Add a label to the GetOutOfRangePersonalityDescription test

 Bugs:
 * Fix a few more minor documentation spelling errors found by codespell
 * Point to the main site rather than the wiki for RDM test help
 * Fix the build on FreeBSD 10.0, 10.4 and 11
 * Make PID store loading more robust
 * Clarify some of the man page options for olad
 * Fix some dead links in documentation
 * Properly close socket on destruction of Python OlaClient
 * Fix various Python 3 compatibility issues
 * Try and fetch index and type via sockaddr_dl struct for better *BSD support
 * Backport some Debian changes
 * Use C++11 if required by random #1477
 * Correct uartdmx plugin documentation for the example UART #1320
 * Ensure we don't return unsuitable network indices via RDM #1478
 * Ensure a non-existent UID exits with a non-zero exit code
 * Fix a bug in how we simulate RDM responses and collisions
 * Fix an RDM discovery bug for how we handle phantom UIDs
 * Ensure _close_callback is assigned before we might need to run it in the
   Python API
 * Fix a bug with the Python ola_rdm_get and PROXIED_DEVICES display
 * Renamed RESONSE_INVALID_DESTINATION(sic) to RESPONSE_INVALID_DESTINATION in
   the ArduinoWidget code
 * Fix compatibility with GCC 9 #1553 Debian #925793
 * Ensure the GPIO plugin correctly writes to the last configured pin
 * Fix compatibility with ncurses 6
 * Fix compatibility with Protobuf 3.7 and newer (tested with up to 3.12.2)
   #1550, #1634
 * Rename CircularDepdendancyException(sic) to CircularDependencyException in
   the Python RDM Test code, also the relevant comments
 * Fix a minor longstanding logging bug with Avahi subtype registration errors
 * Fix Enttec USB Pro Mk2 firmware >=4.15 RDM Discovery #1631
 * Fix compatibility with libmicrohttpd v0.9.71 and newer #1650
 * Fix some Doxygen escaping
 * Rename AppendMultipler(sic) to AppendMultiplier in the RDM messaging code
 * Fix a few edge cases in ola_trigger around whitespace in the config files
 * Fix compatibility with clang 7 and newer (tested with up to 11) #1564

 Internal:
 * Replace "readdir_r" with "readdir" since the former has been deprecated
 * Add a unit test for the functions that used readdir_r before
 * Add a new 'make flake8' target #1619
 * Add a new 'make cpplint' target
 * Add a new 'make lint' target which runs the flake8 and cpplint runs

13/7/2018 ola-0.10.7
 Features:
 * Allow multiple KiNet ports to be patched to the same universe #1414
 * Allow multiple FTDI ports to be patched to the same universe
 * Support libftdi1 as well as libftdi0 in FTDI DMX #1012 Debian #810374
 * Support FTDI DMX via libftdi1 on Windows

 RDM Tests:
 * Fix a bug in the RDM tests when a responder supports sensor PIDs #1388
 * Fix a bug in imports for rdm_responder_test.py

 Bugs:
 * Fix the build on Windows
 * Entire codebase now passes codespell testing
 * Fix a Python 3 bug in the API
 * Fix a few longstanding bugs in the Python ola_rdm_discover
 * Correct the OSC config option name for output format in the documentation
 * Use top_builddir in more places in the build
 * Use libusb_set_option where available
 * Fix the udev rules
 * Downstream Debian fixes
 * Fix an incorrect response if we have the sensor PIDs but no sensors
 * Stop trying to build opendmx kernel plugin on Windows
 * Prefer Avahi over Bonjour for our DiscoveryAgent, in case the Bonjour
   version is actually just Avahi's compatibility layer
 * Clean up the libolaserver pkg-config file

 Internal:
 * Various minor log formatting tidying up

07/01/2018 ola-0.10.6

python/examples/ola_rdm_get.py
#!/usr/bin/env python
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# ola_rdm_get.py
# Copyright (C) 2010 Simon Newton

'''Get a PID from a UID.'''

from __future__ import print_function

import cmd
import getopt
import os.path
import readline
import sys
import textwrap

from ola.ClientWrapper import ClientWrapper
from ola.OlaClient import OlaClient, RDMNack
from ola.RDMAPI import RDMAPI
from ola.UID import UID

from ola import PidStore

__author__ = 'nomis52@gmail.com (Simon Newton)'


def Usage():
  print(textwrap.dedent("""\
  Usage: ola_rdm_get.py --universe <universe> --uid <uid> <pid>

  Get the value of a pid for a device.
  Use 'ola_rdm_get --list-pids' to get a list of pids.

    -d, --sub-device <device> target a particular sub device (default is 0)
    -h, --help                display this help message and exit.
    -i, --interactive         interactive mode
    -l, --list-pids           display a list of pids
    -p, --pid-location        the directory to read PID definitions from
    --uid                     the UID to send to
    -u, --universe <universe> Universe number."""))


wrapper = None


class ResponsePrinter(object):
  def __init__(self):
    self.pid_store = PidStore.GetStore()
    # we can provide custom printers for PIDs below
    self._handlers = {
        'SUPPORTED_PARAMETERS': self.SupportedParameters,
        'PROXIED_DEVICES': self.ProxiedDevices,
    }

  def PrintResponse(self, uid, pid_value, response_data):
    pid = self.pid_store.GetPid(pid_value, uid.manufacturer_id)
    if pid is None:
      print('PID: 0x%04hx' % pid_value)
      self.Default(uid, response_data)
    else:
      print('PID: %s' % pid)
      if pid.name in self._handlers:
        self._handlers[pid.name](uid, response_data)
      else:
        self.Default(uid, response_data)

  def SupportedParameters(self, uid, response_data):
    params = sorted([p['param_id'] for p in response_data['params']])
    for pid_value in params:
      pid = self.pid_store.GetPid(pid_value, uid.manufacturer_id)
      if pid:
        print(pid)
      else:
        print('0x%hx' % pid_value)

  def ProxiedDevices(self, uid, response_data):
    uids = []
    for uid_data in response_data['uids']:
      uids.append(uid_data['uid'])
    for uid in sorted(uids):
      print(uid)

  def Default(self, uid, response_data):
    if isinstance(response_data, dict):
      for key, value in response_data.items():
        print('%s: %r' % (key, value))
    else:
      print(response_data)


class InteractiveModeController(cmd.Cmd):
  """Interactive mode!"""
  def __init__(self, universe, uid, sub_device, pid_location):
    """Create a new InteractiveModeController.

    Args:
      universe:
      uid:
      sub_device:
      pid_location:
    """
    cmd.Cmd.__init__(self)
    self._universe = universe
    self._uid = uid
    self._sub_device = sub_device
    self.pid_store = PidStore.GetStore(pid_location)
    self.wrapper = ClientWrapper()
    self.client = self.wrapper.Client()
    self.rdm_api = RDMAPI(self.client, self.pid_store)
    self._uids = []
    self._response_printer = ResponsePrinter()

    # tuple of (sub_device, command_class, pid)
    self._outstanding_request = None

    self.prompt = '> '

  def emptyline(self):
    pass

  def do_exit(self, s):
    """Exit the interpreter."""
    return True

  def do_EOF(self, s):
    print('')
    return self.do_exit('')

  def do_uid(self, line):
    """Sets the active UID."""
    args = line.split()
    if len(args) != 1:
      print('*** Requires a single UID argument')
      return

    uid = UID.FromString(args[0])
    if uid is None:
      print('*** Invalid UID')
      return

    if uid not in self._uids:
      print('*** UID not found')
      return

    self._uid = uid
    print('Fetching queued messages...')
    self._FetchQueuedMessages()
    self.wrapper.Run()

  def complete_uid(self, text, line, start_index, end_index):
    tokens = line.split()
    if len(tokens) > 1 and text == '':
      return []

    uids = [str(uid) for uid in self._uids if str(uid).startswith(text)]
    return uids

  def do_subdevice(self, line):
    """Sets the sub device."""
    args = line.split()
    if len(args) != 1:
      print('*** Requires a single int argument')
      return

    try:
      sub_device = int(args[0])
    except ValueError:
      print('*** Requires a single int argument')
      return

    if sub_device < 0 or sub_device > PidStore.ALL_SUB_DEVICES:
      print('*** Argument must be between 0 and 0x%hx' %
            PidStore.ALL_SUB_DEVICES)
      return
    self._sub_device = sub_device

  def do_print(self, line):
    """Prints the current universe, UID and sub device."""
    print(textwrap.dedent("""\
      Universe: %d
      UID: %s
      Sub Device: %d""" % (
        self._universe,
        self._uid,
        self._sub_device)))

  def do_uids(self, line):
    """List the UIDs for this universe."""
    self.client.FetchUIDList(self._universe, self._DisplayUids)
    self.wrapper.Run()

  def _DisplayUids(self, state, uids):
    self._uids = []
    if state.Succeeded():
      self._UpdateUids(uids)
      for uid in uids:
        print(str(uid))
    self.wrapper.Stop()

  def do_full_discovery(self, line):
    """Run full RDM discovery for this universe."""
    self.client.RunRDMDiscovery(self._universe, True, self._DiscoveryDone)
    self.wrapper.Run()

  def do_incremental_discovery(self, line):
    """Run incremental RDM discovery for this universe."""
    self.client.RunRDMDiscovery(self._universe, False, self._DiscoveryDone)
    self.wrapper.Run()

  def _DiscoveryDone(self, state, uids):
    if state.Succeeded():
      self._UpdateUids(uids)
    self.wrapper.Stop()

  def _UpdateUids(self, uids):
    self._uids = []
    for uid in uids:
      self._uids.append(uid)

  def do_list(self, line):
    """List the pids available."""
    names = []
    for pid in self.pid_store.Pids():
      names.append('%s (0x%04hx)' % (pid.name.lower(), pid.value))
    if self._uid:
      for pid in self.pid_store.ManufacturerPids(self._uid.manufacturer_id):
        names.append('%s (0x%04hx)' % (pid.name.lower(), pid.value))
    names.sort()
    print('\n'.join(names))

  def do_queued(self, line):
    """Fetch all the queued messages."""
    self._FetchQueuedMessages()
    self.wrapper.Run()

  def do_get(self, line):
    """Send a GET command."""
    self.GetOrSet(PidStore.RDM_GET, line)

  def complete_get(self, text, line, start_index, end_index):
    return self.CompleteGetOrSet(PidStore.RDM_GET, text, line)

  def do_set(self, line):
    """Send a SET command."""
    self.GetOrSet(PidStore.RDM_SET, line)

  def complete_set(self, text, line, start_index, end_index):
    return self.CompleteGetOrSet(PidStore.RDM_SET, text, line)

  def CompleteGetOrSet(self, request_type, text, line):
    if len(line.split(' ')) > 2:
      return []
    pids = [pid for pid in self.pid_store.Pids()
            if pid.name.lower().startswith(text)]
    if self._uid:
      for pid in self.pid_store.ManufacturerPids(self._uid.manufacturer_id):
        if pid.name.lower().startswith(text):
          pids.append(pid)

    # now check if this type of request is supported
    pid_names = sorted([pid.name.lower() for pid in pids
                        if pid.RequestSupported(request_type)])

    return pid_names

  def GetOrSet(self, request_type, line):
    if self._uid is None:
      print('*** No UID selected, use the uid command')
      return

    args = line.split()
    command = 'get'
    if request_type == PidStore.RDM_SET:
      command = 'set'
    if len(args) < 1:
      print('%s <pid> [args]' % command)
      return

    pid = None
    try:
      pid = self.pid_store.GetPid(int(args[0], 0), self._uid.manufacturer_id)
    except ValueError:
      pid = self.pid_store.GetName(args[0].upper(), self._uid.manufacturer_id)
    if pid is None:
      print('*** Unknown pid %s' % args[0])
      return

    if not pid.RequestSupported(request_type):
      print('*** PID does not support command')
      return

    if request_type == PidStore.RDM_SET:
      method = self.rdm_api.Set
    else:
      method = self.rdm_api.Get

    try:
      if method(self._universe,
                self._uid,
                self._sub_device,
                pid,
                self._RDMRequestComplete,
                args[1:]):
        self._outstanding_request = (self._sub_device, request_type, pid)
        self.wrapper.Run()
    except PidStore.ArgsValidationError as e:
      args, help_string = pid.GetRequestDescription(request_type)
      print('Usage: %s %s %s' % (command, pid.name.lower(), args))
      print(help_string)
      print('')
      print('*** %s' % e)
      return

  def _FetchQueuedMessages(self):
    """Fetch messages until the queue is empty."""
    pid = self.pid_store.GetName('QUEUED_MESSAGE', self._uid.manufacturer_id)
    self.rdm_api.Get(self._universe,
                     self._uid,
                     PidStore.ROOT_DEVICE,
                     pid,
                     self._QueuedMessageComplete,
                     ['error'])

  def _RDMRequestComplete(self, response, unpacked_data, unpack_exception):
    if not self._CheckForAckOrNack(response):
      return

    # at this stage the response is either a ack or nack
    self._outstanding_request = None
    self.wrapper.Stop()

    if response.response_type == OlaClient.RDM_NACK_REASON:
      print('Got nack with reason: %s' % response.nack_reason)
    elif unpack_exception:
      print(unpack_exception)
    else:
      self._response_printer.PrintResponse(self._uid, response.pid,
                                           unpacked_data)

    if response.queued_messages:
      print('%d queued messages remain' % response.queued_messages)

  def _QueuedMessageComplete(self, response, unpacked_data, unpack_exception):
    if not self._CheckForAckOrNack(response):
      return

    if response.response_type == OlaClient.RDM_NACK_REASON:
      if (self._outstanding_request and
          response.sub_device == self._outstanding_request[0] and
          response.command_class == self._outstanding_request[1] and
          response.pid == self._outstanding_request[2].value):
        # we found what we were looking for
        self._outstanding_request = None
        self.wrapper.StopIfNoEvents()

      elif (response.nack_reason == RDMNack.NR_UNKNOWN_PID and
            response.pid == self.pid_store.GetName('QUEUED_MESSAGE').value):
        print('Device doesn\'t support queued messages')
        self.wrapper.StopIfNoEvents()
      else:
        print('Got nack for 0x%04hx with reason: %s' % (
            response.pid, response.nack_reason))

    elif unpack_exception:
      print('Invalid Param data: %s' % unpack_exception)
    else:
      status_messages_pid = self.pid_store.GetName('STATUS_MESSAGES')
      if (response.pid == status_messages_pid.value and
          unpacked_data.get('messages', []) == []):
        self.wrapper.StopIfNoEvents()
        return

      self._response_printer.PrintResponse(self._uid,
                                           response.pid,
                                           unpacked_data)

    if response.queued_messages:
      self._FetchQueuedMessages()
    else:
      self.wrapper.StopIfNoEvents()

  def _CheckForAckOrNack(self, response):
    """Check for all the different error conditions.

    Returns:
      True if this response was an ACK or NACK, False for all other cases.
    """
    if not response.status.Succeeded():
      print(response.status.message)
      self.wrapper.Stop()
      return False

    if response.response_code != OlaClient.RDM_COMPLETED_OK:
      print(response.ResponseCodeAsString())
      self.wrapper.Stop()
      return False

    if response.response_type == OlaClient.RDM_ACK_TIMER:
      # schedule the fetch
      self.wrapper.AddEvent(response.ack_timer, self._FetchQueuedMessages)
      return False

    return True


def main():
  readline.set_completer_delims(' \t')
  try:
    opts, args = getopt.getopt(sys.argv[1:], 'd:hilp:u:',
                               ['sub-device=', 'help', 'interactive',
                                'list-pids', 'pid-location=', 'uid=',
                                'universe='])
  except getopt.GetoptError as err:
    print(str(err))
    Usage()
    sys.exit(2)

  universe = None
  uid = None
  sub_device = 0
  list_pids = False
  pid_location = None
  interactive_mode = False
  for o, a in opts:
    if o in ('-d', '--sub-device'):
      sub_device = int(a)
    elif o in ('-h', '--help'):
      Usage()
      sys.exit()
    elif o in ('-i', '--interactive'):
      interactive_mode = True
    elif o in ('-l', '--list-pids'):
      list_pids = True
    elif o in ('--uid',):
      uid = UID.FromString(a)
    elif o in ('-p', '--pid-location',):
      pid_location = a
    elif o in ('-u', '--universe'):
      universe = int(a)

  if universe is None and not list_pids:
    Usage()
    sys.exit()

  if not uid and not list_pids and not interactive_mode:
    Usage()
    sys.exit()

  # try to load the PID store so we fail early if we're missing PIDs
  try:
    PidStore.GetStore(pid_location)
  except PidStore.MissingPLASAPIDs as e:
    print(e)
    sys.exit()

  controller = InteractiveModeController(universe,
                                         uid,
                                         sub_device,
                                         pid_location)
  if interactive_mode:
    sys.stdout.write('Available Uids:\n')
    controller.onecmd('uids')
    controller.cmdloop()
    sys.exit()

  if list_pids:
    controller.onecmd('list')
    sys.exit()

  if len(args) == 0:
    Usage()
    sys.exit()

  request_type = 'get'
  if os.path.basename(sys.argv[0]) == 'ola_rdm_set.py':
    request_type = 'set'
  controller.onecmd('%s %s' % (request_type, ' '.join(args)))


if __name__ == '__main__':
  main()
